enhancements/microshift/dummy-gateway-interface-ovn.md

```diff
@@ -0,0 +1,458 @@
+
+On the Microshift side, we will need to make the following change:
+1. We need to update the microshift-ovs-init.service to 
+   * Create a dummy interface on the node.
+   * Create br-ex bridge with this dummy interface, and assign a static IP to
+   br-ex. 
+   * Since we don't need NetworkManager to manage the DHCP client. We can delete
+   those logic from the configure-ovs.sh.
+   * For upgrading, we also need to remove the NetworkManager connections
+   created by configure-ovs.sh in previous version.
+2. Run ovnkube-node with flag `--gateway-nexthop=169.254.169.4`.
+
+### Risks and Mitigations
+
+The OVN-Kubernetes configuration described in this document is not used by any
+other known platforms. Therefore there is no CI with this setup in
+OVN-Kubernetes upstream. In order to minimize the risk of regression, we need to
+improve the OVN-Kubernetes CI to cover this setup.
+
+### Drawbacks
+
+This change will break a service if it is `internalTrafficPolicy=Local` and
+backed by hostNetwork pods.
+
+The reason is that, for services with `internalTrafficPolicy=Local`, ovn-k will
+only put local host endpoints to the node switch loadbalancer. If the pod is
+hostNetwork, the local endpoint means the pod with the NodeIP. Normally, the
+ovn-k will use the nodeIP as the ovn-k external IP in node annotation
+`k8s.ovn.org/l3-gateway-config`. However, in our case, as we use a dummy
+interface for br-ex, we have different IPs for nodeIP and ovn-k external IP. So
+ovn-k cannot identify the right local endpoint in such a case. It then put no
+endpoint to the loadbalancer.
+
+However, this problem is not so critical to Microshift. Since we only get one
+node in the cluster. Creating a service with `internalTrafficPolicy=Local` makes
+no sense in such a case. We can fix this problem later until we find a proper
+solution.
+
+## Design Details
+
+### Open Questions [optional]
+
+1. We'll need to allocate a IP address to br-ex bridge. This IP address will not
+   be exposed externally. Shall we use a fixed IP address, or we allow user to
+   be able to specify it?
+
+   A: We decide to reuse the `advertise-address` of kubernetes API server by add
+   this address to br-ex. Users can specify the address with the config flag
+   `apiServer.advertiseAddress` in microshift. The default value is `10.44.0.0`.
+
+### Test Plan
+
+The current tests we have are sufficient for testing this networking change.
+
+### Graduation Criteria
+
+N/A
+
+#### Dev Preview -> Tech Preview
+
+N/A
+
+#### Tech Preview -> GA
+
+N/A
+
+#### Removing a deprecated feature
+
+N/A
+
+### Upgrade / Downgrade Strategy
+
+When doing upgrade, the microshift-ovs-init.service will be responsible for
+deleting the br-ex bridge and NetworkManager config files that is created by
+previous version with NetworkManager.
+
+Downgrade cannot be done automatically, users will need to manually remove the
+dummy interface and br-ex bridge created by microshift-ovs-init.service.
```

Comment:
Will this limitation prevent us from testing the upgrade and rollback work? Maybe we can work around that in CI?

/cc @pmtk

Reply:
we dont support downgrade/rollback in OCP. Not sure if microshift has some exception, but overall downgrade has no support without manual steps for OVNK.

Reply:
MicroShift is integrating with the rpm-ostree features for whole-system upgrade/rollback. See #1387 for details. I would expect, based on the way OVNK works in MicroShift, that the system would rebuild the OVN database after the host reboots into the old system. Is that right @zshi-redhat or @pliurh ?

Reply:
> MicroShift is integrating with the rpm-ostree features for whole-system upgrade/rollback. See #1387 for details. I would expect, based on the way OVNK works in MicroShift, that the system would rebuild the OVN database after the host reboots into the old system. Is that right @zshi-redhat or @pliurh ?
> 

Yes. @trozet we run ovn dbs in container in microshift, and upgrade/downgrade in microshift allows host reboot, which then restarts ovnkube containers and rebuilds the ovn dbs. This works slightly different compared with OCP. Would you agree that downgrade in microshift can be supported as described?

Reply:
It seems OK I guess in the context of microshift and a single node. I guess although small, there could be some risk with leftover configuration done as far as like node or pod annotations that a previous version would be incompatible with. We dont test anything like this in our CI so perhaps we need a downgrade job in OCP to ensure this works and doesn't break.

Reply:
@pliurh could you help to capture the work Tim described in the doc that we need a downgrade job in OCP?

Reply:
The downgrade support is orthogonal to this enhancement. The downgrade chapter of this document is to describe the downgrade from a version that has the gateway bridge change to the one that doesn't. So I'll create a new story under [USHIFT-518](https://issues.redhat.com/browse/USHIFT-518) to track the CI work we need to do for ovn-k downgrade.